### ISSUE TITLE
Slow compilation to PDF if a lot of tables are present

### ISSUE THREAD

**Issue by soumilkanwal80** [typst/typst#4560]
### Description

Currently we have an invoice generation system which generates PDF by converting from HTML. I was looking it to migrate to typst for PDF generation.
Issue is our invoice is 8 column with maximum of 0.1M rows.

I tried replicating our template on typst, which I was able to do so.
But on generating PDF it takes a long time(took more than 1 hour and then computer crashed on M3 Pro 16 inch MacBook).

Page dimensions are A4.

I tried both on typst.app and install typst compiler using home-brew

```
#set page("a4")
#set text(font: "DM Sans", fill: rgb("5B616EFF"), weight: 500, size: 10pt)

#show figure: set block(breakable: true)
#figure(
table(
  columns: 8,
  align: center,
  ..for x in range(1, 100000) {
    ([#x], [#(2*x)], [#(3*x)], [#(4*x)], [#(5*x)], [#(6*x)], [#(7*x)], [#(8*x)])
  }
)
)

```

### Operating system

macOS

### Typst version

- [X] I am using the latest version of Typst

**Comment by Enivex** [comment]
Are you saying that performance is poor if 100 is swapped out for 100,000?

**Comment by Enivex** [comment]
With 100,000 it compiles for me locally in 36 seconds. Though it peaks at 18.1 GB memory. I suspect that your computer does not have enough memory to complete the compilation without dipping into the page file. (The web app is a non-starter, since it's limited to 4 GB)

![image](https://github.com/user-attachments/assets/faced27b-5708-4570-b30c-23f90b21084c)
![image](https://github.com/user-attachments/assets/a5269c4a-0f94-449f-ad90-f9c6a17a74af)

**Comment by soumilkanwal80** [comment]
@Enivex my bad its 100000, 
Thanks for the reply I was able to run this after closing and rebooting everything on my Mac.

My workload is similar to this script, where we have a bunch of smaller tables each having 1-10 rows but totall rows are around 0.1M

File is very large around 12k lines so couldn't post code snippet here
Github link: https://github.com/soumilkanwal80/Typst-Exp/blob/main/multi_table.typ

**Comment by soumilkanwal80** [comment]
The biggest difference I have observed is in the first case memory grows quickly, but in this case it allocates very slowly, I am guessing it tries to render each table one by one and therefore taking a lot of time?

**Comment by Daaiid** [comment]
I started investigating this issue with the given file using flamegraphs and I found the hotspot:
* About 80% of the time is spent in the function `typst_library::foundations::styles::Styles::apply`
* The common call site is at `typst_realize::prepare`
* `prepare` is called at (and assumably inlined so as to not show up on the flamegraph) in `typst_realize::visit_show_rules`
* The rest of the call graph is a stack of recursive calls to `typst_realize::visit` and `typst_realize::visit_styled` which indicates the visitor pattern being used to traverse the document hierarchy.

My guess to what the root of the problem is:
The problematic `apply` function seems to append `EcoVec` to each other. [The crate `ecow`](https://docs.rs/ecow/latest/ecow/) implementing this type describes it as a "clone on write vector". I image this causes *a lot* of copies of the vector being created.

I don't know the code base or Rust well enough to come up with a reasonable fix to this problem for now.
My hope is this initial analysis will help someone else get started with this issue.  

Here is the flamegraph:
![Image](https://github.com/user-attachments/assets/606c6dc8-7c02-445e-b3d5-a0bc51a6c632)

**Comment by thuiop** [comment]
This is fairly old so it may not be that useful but I looked a bit into it. I am not reaching the same result as the previous commenter with the flamegraph. From what I see, the compilation is dominated by the actual rendering of the table. An interesting thing is that it is split in three parts: measuring the rows, measuring the columns, and actually rendering the rows. A given cell may be rendered in each of these steps. This is supposed to be mostly memoized (through the `layout_fragment` function) but the code does not seem to fully leverage that as it spends a lot of time in all three of the steps. This may give a way to speed this up. In the meantime, a workaround that can be implemented in your Typst code is specifying the width and height of the cells manually through the `columns` and `rows` arguments of `table`. I am getting a x2 speedup by doing this. The remainder of the work is to render the content of the cells itself, so there probably is not any particular gain to make here.

<img width="2414" height="635" alt="Image" src="https://github.com/user-attachments/assets/341a13fa-75a3-4256-b265-69625b2534fe" />

**Comment by jqwv** [comment]
I think one problem here is that tables in typst are currently not slim and are created [from as many small elements as possible](https://github.com/typst/typst/issues/6896#issuecomment-3281292141). I assume that rendering even a single table is very costly in typst.
Thus, a workaround would be to define a table template that can be copied in the document at some place with different values instead of telling typst to rerender the full thing again and again. This is even possible at pdf level for example by reserving a specific amounts of bytes and later copying in the byte patch at this place in the PDF file.
Of course this is not the answer you want to hear :D but it would be O(N) or even O(1) runtime.

**Comment by laurmaedje** [comment]
I doubt that the number of lines is the problem. Drawing the lines happens at the very end of drawing the table. Looking at the flame graphs above, the expensive part is measuring and laying out the cells themselves.

**Comment by jqwv** [comment]
Yes, makes sense. Still, the proposed workaround would make it really fast but is a bit technical. It is even possible to generate simple pdf tables using a handwritten external pdf table tool and then load the pdf as an `image`.

**Comment by thuiop** [comment]
Yes, I believe Laurenz is correct. Changing the code to
```
#table(
  columns: (1cm,1cm,1cm,1cm,1cm,1cm,1cm,1cm),
  align: center,
  ..for x in range(1, 100000) {
    ([#x], [#(2*x)], [#(3*x)], [#(4*x)], [#(5*x)], [#(6*x)], [#(7*x)], [#(8*x)])
  }
)
```
for instance (you can do a similar things with rows) already reduces the compilation time significantly. For 5000 rows, specifying the columns width makes it go from 1.45s to 1.05s. Further specifying the row height makes it go to 670ms. Adding `stroke: none` only drops it further to 620ms, so drawing the actual lines does not seem to be the main blocker here. And I am pretty sure the remaining time is actually spent in rendering the content in each cell. I do not think there really is a way around this.

The part where it could in principle be improved is by reusing the rendering done when measuring the columns and rows for the actual rendering of the cells. The code superficially looks like it should be doing that, but evidently it does not; there likely is some subtlety here I am not grasping (I would guess that you need to rerender the cell with the proper height and width in the end? But there could be some reusing when you have auto rows AND columns).

**Comment by mayankguptadotcom** [comment]
So - I did what the contribution guideline says not to do - vibecode it. Reason - I'm not a programmer, but I wanted to see if this can be fixed. 

`CONTRIBUTING.md`, step 2:

> "Do not vibecode the change! Contributions that were implemented by an AI model will not be accepted."

Using AI - I built a Drupal Module to generate the PDF using typst binary and while benchmarking, against a very large table - my container ran out of memory.

Below are the results after the fix done by Claude Opus 5

**The results** (measured in a Linux container, same document, three runs each):

| table size | before | after | saved |
|---|---|---|---|
| 250 rows | 175 MB | 95 MB | 45% |
| 2,000 rows | 1,198 MB | 520 MB | 57% |
| 10,000 rows | 5,750 MB | 2,372 MB | **59%** |

The PDF that comes out is **exactly the same file**, byte for byte, in every case. Typst's own test suite — 3,684 tests — still passes.

**What it costs:** the compile is about **7.5% slower**. 

I would've raised the request, but didn't as per the contribution guidelines - but here's the patch and instructions, in case someone wants it.

[typst-table-memory-patch.tar.gz](https://github.com/user-attachments/files/30810457/typst-table-memory-patch.tar.gz)

[APPLYING.md](https://github.com/user-attachments/files/30810492/APPLYING.md)

**Comment by thuiop** [comment]
As one would expect from a non-programmer using AI, this proposed patch is extremely misguided. It just fails to understand that the memoization is there mostly for `typst watch` and not that much for reuse of elements across the documents (although it supports that of course). Thus, the suggestion it makes only makes sense for one-shot compiles and completely ignore `typst watch`; if you have a big table it will recompile the whole document every time, even if you did not change the table. Surely there is something smart to do with the memoization that would allow `typst compile` to use less RAM (and gain a bit of wall clock time), as talked about in #6508, but this definitely is not it.

**Comment by mayankguptadotcom** [comment]
Thank you for your comment and guidance. Fair point on `watch` — I didn't know about it. The benchmarks were all one-shot compiles, so they can't speak to the case the memoization actually exists for. Thanks for [#6508](https://github.com/typst/typst/issues/6508), that looks like the right thread to follow.

That's why I do this — to learn, with the intent to contribute. I'll get things wrong, and this was one of them.

I appreciate you taking the time to engage. It's made me want to understand this more deeply than I do.
